CORE-6287: Simulating signing service, verification

LizR3 opened the pull request:
A key store is passed into each node which will generate a public key on request. Only RSA keys are generated, regardless of spec. The keys are made available for each member through the MemberLookup.

The signing service fakes signing by wrapping the clear data, PEM-encoded key and other parameters in a JSON wrapper.

The verification service parses the wrapper and checks the key, signature spec and clear data against the expected values.

No encryption is ever actually performed. A prominent warning has been added to the README.

LizR3 committed:
CORE-6287: Simulating signing service, verification

mnesbit commented:
Note we should always favour ECDSA, or EdDSA over RSA. RSA is not a good statement on our using current cryptography.

LizR3 commented:
@mnesbit Deliberately used RSA to avoid the appearance of actually doing any cryptography (nothing is encrypted!), but I can change it for sure.

alexey-kadyrov-r3 commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/cordapptestutils/internal/signing/KeyStore.kt:
```diff
@@ -0,0 +1,9 @@
+import net.corda.cordapptestutils.crypto.HsmCategory
+import java.security.PublicKey
+
+interface KeyStore {
```
Don't use `KeyStore` name as it's confusing due that the name is the well defined class in JAVA

LizR3 replied:
@alexey-kadyrov-r3 Would `SimKeyStore` work, or is that still too similar?

alexey-kadyrov-r3 replied:
Yep

LizR3 committed:
CORE-6287: Change key algorithm, KeyStore name

LizR3 committed:
CORE-6287: Fixed strangely failing test

IntelliJ will quite happily refactor "Unit" as a return type, but it messes with detecting any annotations you put on the method. Also better logging.

mnesbit commented on cordapp-test-utils/example-app/src/main/kotlin/net/cordacon/example/AbsenceCallResponderFlow.kt:
```diff
@@ -6,15 +6,24 @@ import net.corda.v5.application.flows.InitiatedBy
 @InitiatedBy("absence-call")
 class AbsenceCallResponderFlow: ResponderFlow {
 
+    private companion object {
```
We will have to watch the use of companion objects in Corda 5! They are not in fact reliable in the hot-hot HA model that we operate. Logger are special case handled by the platform, but one must never store any state in the companion object

mnesbit commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/cordapptestutils/internal/signing/BaseSimKeyStore.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    override fun generateKey(alias: String, hsmCategory: HsmCategory, scheme: String) : PublicKey {
+        val key = keyGenerator.generateKeyPair().public
```
I am not totally sure why we would throw away the PrivateKey at this point? It seems simplest to my mind to represent the crypto with actual keys and signatures.

mnesbit reviewed (approved):
I will approve so that things move along, but I am not quite sure at some of the design choices here

ronanbrowne reviewed (approved):
LGTM ( Build change reviewed )

alexey-kadyrov-r3 reviewed (approved):
LGTM
